More PoS pseudocode edits

brentstone opened the pull request:
Missed some last time around. Leaving on draft again until I think I've found everything.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -240,10 +248,12 @@ func bond(validator_address, delegator_address, amount)
+    update_voting_power(validator_address, pipeline_length)
+    update_total_voting_power(pipeline_length)
+    // Only update validator sets if the validator is not jailed, since a jailed validator is not in a validator set
+    if !is_jailed(validator_address) then
```
This is not needed, we remove a validator from active and inactive sets when jailing it.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -292,10 +302,14 @@ func unbond(validator_address, delegator_address, total_amount)
       update_voting_power(validator_address, pipeline_length)
       update_total_voting_power(pipeline_length)
-      update_validator_sets(validator_address, pipeline_length)
+      if !is_jailed(validator_address) then
```
Same as above: not needed.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -320,15 +334,16 @@ func withdraw(validator_address, delegator_address)
+  forall (slash in set_slashes in slash.epoch order) do
+    //Update amount with slashes that happened more than `unbonding_length` before this slash
+    forall (slashed_amount in computed_amounts s.t. slashed_amount.epoch + unbonding_length < slash.epoch) do
+      updated_amount = updated_amount.checked_sub(slashed_amount.amount)
```
I would rather keep language specifics out. `checked_sub` is undefined within the spec. I guess you want to avoid `updated_amount` going negative. Maybe the more intuitive way of doing this is to simply break the loop if `update_amount` becomes negative and return 0 in that case.

brentstone replied:
Addressed now in #50

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -292,10 +302,14 @@ func unbond(validator_address, delegator_address, total_amount)
+      
+      var pipeline_stake = read_epoched_field(validators[validator_address].total_deltas, cur_epoch + pipeline_length, 0)
+      var token_change = min{ amount_after_slashing, pipeline_stake }
+      update_total_deltas(validator_address, pipeline_length, -1*token_change)
```
It looks to me that it would be enough to modify `compute_amount_after_slashing` to always return 0 or greater. If it returns 0, it means that the unbond is fully slashed. I believe the above lines are then not needed.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -510,19 +524,21 @@ end_of_epoch()
 
   var set_validators = {val | val = slash.validator && slash in enqueued_slashes[cur_epoch]}
   forall (validator_address in set_validators) do
+    var total_staked = read_epoched_field(validators[validator_address].total_deltas, slash.epoch, 0)
```
`slash` is undefined here, i.e, `slash.epoch` is as well. What's the motivation to move it out of the forall loop of slashes?

brentstone replied:
I've changed `end_of_epoch()` again now in 61c733f to correct what I had done before and I think to solve the issue of `total_unbonded > total_staked`

angbrav reviewed (changes requested):
Thanks @brentstone! Everything looks good to me but the part that computes the final slash rate. We need to change `compute_final_rate` to implement what's described in https://specs.namada.net/economics/proof-of-stake/cubic-slashing.html. To mention one discrepancy, The pseudocode computes `slash.voting_power_fraction` (which is undefined and I assume that it is `stake_fraction`) by multiplying the misbehaving validator's stake at the misbehaving epoch to the infraction type rate. The spec does something different by dividing the validator's voting power by the total voting power (total PoS voting power)

Once we fix this, which should be easy, this PR can be merged.

brentstone commented:
> Thanks @brentstone! Everything looks good to me but the part that computes the final slash rate. We need to change `compute_final_rate` to implement what's described in https://specs.namada.net/economics/proof-of-stake/cubic-slashing.html. To mention one discrepancy, The pseudocode computes `slash.voting_power_fraction` (which is undefined and I assume that it is `stake_fraction`) by multiplying the misbehaving validator's stake at the misbehaving epoch to the infraction type rate. The spec does something different by dividing the validator's voting power by the total voting power (total PoS voting power)
> 
> Once we fix this, which should be easy, this PR can be merged.

Thanks for the note, I've addressed this in 48c13ba6e4475528111c100e980c50a9647f2c92. I've done some refactoring in order to sync this with the Namada spec and codebase. In particular, the `Slash` object has a field `voting_power` rather than `stake_fraction`, which indicates the voting power used to commit the infraction. As such, the function `compute_stake_fraction` is no longer needed and removed. I've also renamed `compute_final_rate` to `compute_cubic_rate`, as the final rate is computed later for each within the loop.

Lmk your thoughts.

angbrav commented:
I am still confused by the pseudocode in Namda's spec:
- Namada's spec computes the final rate as `min{1, max{get_min_slash_rate(slash), 9 * voting_power_fraction^2}}` (adapted to pseudocode model notation) and the pseudocode model does `max{get_min_slash_rate(slash), min{1, 9 * voting_power_fraction^2}}`, which is not equivalent. But I think there is something wrong in Namada's spec as `slash` is undefined. Is then what's currently in the pseudocode model the correct spec?
- It is unclear at which epoch the misbehaving validator's stake is retrieved in `validator.voting_power` by the Namada's spec. This is clearly underspecified in that pseudocode. The pseudocode model currently divides the validator's voting power at the misbehaving epoch (retrieved when the evidence is processed) by the total voting power at the current epoch. I am not sure that's intended. If it is, what's the intuition behind this?

Some minor thing. I would do something about
```go
func get_min_slash_rate(infraction){
  switch infraction
    case duplicate_vote: return duplicate_vote_rate
    case light_client_attack: return light_client_attack
    default: panic()
}
```
and add a field to the slash record with type of infraction for instance such that we call `get_min_slash_rate` for a `slash` by `get_min_slash_rate(slash.type)` instead of `get_min_slash_rate(slash)`.

brentstone commented:
> I am still confused by the pseudocode in Namda's spec:
> 
> * Namada's spec computes the final rate as `min{1, max{get_min_slash_rate(slash), 9 * voting_power_fraction^2}}` (adapted to pseudocode model notation) and the pseudocode model does `max{get_min_slash_rate(slash), min{1, 9 * voting_power_fraction^2}}`, which is not equivalent. But I think there is something wrong in Namada's spec as `slash` is undefined. Is then what's currently in the pseudocode model the correct spec?
> * It is unclear at which epoch the misbehaving validator's stake is retrieved in `validator.voting_power` by the Namada's spec. This is clearly underspecified in that pseudocode. The pseudocode model currently divides the validator's voting power at the misbehaving epoch (retrieved when the evidence is processed) by the total voting power at the current epoch. I am not sure that's intended. If it is, what's the intuition behind this?
> 
> Some minor thing. I would do something about
> 
> ```go
> func get_min_slash_rate(infraction){
>   switch infraction
>     case duplicate_vote: return duplicate_vote_rate
>     case light_client_attack: return light_client_attack
>     default: panic()
> }
> ```
> 
> and add a field to the slash record with type of infraction for instance such that we call `get_min_slash_rate` for a `slash` by `get_min_slash_rate(slash.type)` instead of `get_min_slash_rate(slash)`.

Yeah, so first, what is in the pseudocode is more precise in terms of specifying the actual code of Namada. The Namada spec for cubic slashing was also updated today a bit, but nevertheless is a bit more general than an approach that walks through the code very precisely. Anyway, will make some changes to the Namada spec as there are currently some rendering issues and things can be specified more precisely.

You are right that I have a bug in the epochs used. Thanks, I've fixed this. Also have integrated your suggested change regarding `get_min_slash_rate`. See the changes in 6dfa21e.

brentstone commented:
Since you brought up certain things that were undefined as well, I also notice that `evidence` with the fields `validator` and `epoch` may be undefined as well.

angbrav commented:
> Since you brought up certain things that were undefined as well, I also notice that `evidence` with the fields `validator` and `epoch` may be undefined as well.

I would simply assume that `new_evidence` has `type`, `epoch` and `validator` as arguments, instead of the undefined `evidence` struct`.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -345,7 +347,7 @@ func new_evidence(evidence)
   //create slash
   var total_staked = read_epoched_field(validators[evidence.validator].total_deltas, evidence.epoch, 0)
-  var slash = Slash{epoch: evidence.epoch, validator: evidence.validator, rate: 0, stake_fraction: compute_stake_fraction(evidence.type, total_staked)}
+  var slash = Slash{epoch: evidence.epoch, validator: evidence.validator, rate: 0, voting_power: total_staked}
```
See [my comment](https://github.com/informalsystems/partnership-heliax/pull/43#issuecomment-1560660573), it is probably better to remove `evidence` struct, hence my suggestion.
```suggestion
  var slash = Slash{epoch: epoch, validator: validator, rate: 0, type: type, voting_power: total_staked}
```

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -54,7 +54,8 @@ type Slash struct {
   rate float
-  stake_fraction float //new in cubic slashing
+  voting_power VotingPower // the voting power used to commit the infraction
+  slash_type string
```
```suggestion
  type string
```

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -552,13 +554,25 @@ end_of_epoch()
 {
+  var voting_power = 0
+  forall (validator in validator_sets[epoch].active)
+    voting_power += validator.voting_power
```
```suggestion
    voting_power += `read_epoched_field(validators[validator].voting_power, epoch, 0)`
```
I know we are using `validator.voting_power` in some other places, e.g., `update_total_voting_power` or `update_validator_sets`, but I think we should not. We can fix the other functions in a different PR.

angbrav commented on 2022/Q4/artifacts/PoS-pseudocode/PoS-model.md:
```diff
@@ -552,13 +554,25 @@ end_of_epoch()
-//Cubic slashing function
-compute_final_rate(slashes)
+// Get total stake of active (consensus-participating) validators
+get_total_active_voting_power(epoch)
```
It may be useful to precompute both the total active voting power per epoch and the total inactive voting power, as opposed to what we do now: to precomputing the sum of both in `update_total_voting_power`.

brentstone replied:
perhaps this can be in another PR where we make updates as suggested in https://github.com/informalsystems/partnership-heliax/pull/43#discussion_r1203744479?

angbrav reviewed (commented):
I have left a bunch of comments, but nothing serious.

angbrav reviewed (approved):
LGTM, let's merge it, thanks!
